snuba/query/expressions.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Iterator, Optional, Sequence
+
+
+@dataclass(frozen=True)
+class Expression(ABC):
+    """
+    A node in the Query AST. This can be a leaf or an intermediate node.
+    It represents an expression that can be resolved to a value. This
+    includes column names, function calls and boolean conditions (which are
+    function calls themselves in the AST), literals, etc.
+
+    The root of the tree is not a Node itself yet (since it is the Query object).
+    Representing the root as a node itself does not seem very useful right now
+    since we never traverse the full tree. We could revisit that later.
+
+    All expression can have an optional alias.
+    """
+
+    alias: Optional[str]
+
+    @abstractmethod
+    def format(self) -> str:
+        """
+        Turn this node into a string for the Clickhouse query.
+
+        TODO: provide a clickhouse formatter to this method so that, through
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
```

Comment:
I am still trying to figure out the way to format, and I need to do some additional experiments before having a strong opinion. Indeed the idea you are proposing is one of the three approaches:
1) first transform the query into a ClickhosueQuery, that would be structured, and then having a format method on that.
2) the clickhouse module could have a format method that takes a Query, it navigates it and produces the string output. This is simple to understand but the navigation logic becomes complex since the clickhouse query would need to know the internals of the Query object
3) Use a _strategy pattern_, which means passing a ClickhouseFormatter object to the format method in the Query class (which is what I wanted to experiment first). ClickhosueFormatter provides low level primitives (like how to escape columns, how to format a JOIN statement and so on), but the Query object retains the responsibility of traversing the tree, that can be done recursively in a very easy way then. 

Now that we are landing onto the idea that we will have to process a structured clickhouse query as well, I think I should move from 3 to 1.
I am ok in removing the format method for now (anyway it is not implemented at all, it does not change the content of this PR). What do you think ?

Reply:
> Now that we are landing onto the idea that we will have to process a structured clickhouse query as well, I think I should move from 3 to 1. I am ok in removing the format method for now (anyway it is not implemented at all, it does not change the content of this PR). What do you think ?

I agree with both of these points. I think #574 highlighted the need for a separate query object because there are clauses like `PREWHERE` in the ClickHouse query that aren't clauses in the Snuba query. I'd expect that we could reuse a lot of the typing work done here in that class, though.

I think we could figure out how to do one of the other two options if we needed to, but I think it'd be easier to understand (and test) things like prewhere optimization by separating that concern from other ones like formatting.

Reply:
If 3 does not show any evident blocker I am not going to do neither 1 nor 2

Reply:
I'm confused — are we back where we started now? Or something else…?

Reply:
Sorry, I messed up the numbers. Natural numbers are hard, I prefer complex ones.
if **1** does not show any evident blocker I am not going to do neither **2** nor **3**

Reply:
Ok, that makes more sense. 😄